PeerPods CR & controller for handling dangling cloud resources

snir911 opened the pull request:
Fixes: #375 
Here's a suggestion for monitoring and deleting dangling resources, this is RFC as I'd like to get feedback on the general idea and implementation but it's also WIP as there're still things to fix and improve

**Architecture & Workflow:**

Creation time:
With every successful peer-pod VM creation cloud-api-adaptor will create also a PeePod CR (predefined by the operator) which contains the cloud provider and its VM instance id.

Owner references:
The PeerPod CR is owned by the original Pod object that represents the peer-pod, upon pod deletion background cascading deletion getting into action which means the Pod will be deleted first and the GC will handle the owned PeerPod CR later.

Deletion time:
(Normal case) When remote hypervisor will get the stopVM request (upon pod deletion) it will delete the podVM instance and if successful it will remove the finalizer attached to the owned PeerPod object so it can then be cleaned by the GC
(Failure case) In case for some reason cloud-api-adaptor doesn’t honor the delete request of it fails to delete it, the finalizer is not removed and then when peer-pods controller will recognize deletion that the GC is trying to perform for the owned PeerPod object and that it still has the finalizer it will comprehend it needs to preform the deletion of podVM resource based on the PeerPod CR fields

* Currently supports and tested only with libvirt and aws

 TODOs:
- [ ] Docs
- [x] Move to a separate directory?
- [x] Adapt to #710
- [ ] Tests
- [ ] support more providers
- general fixes &  improvements

Depends-on: #755 (caa side)
to test use https://github.com/snir911/cloud-api-adaptor/tree/375merged
with the mentioned images

yoheiueda commented on peer-pod-controller/controllers/peerpod_controller.go:
```diff
@@ -0,0 +1,148 @@
+
+func (r *PeerPodReconciler) setProvider(ctx context.Context, cloudName string) error {
+	r.configsToEnv(cloudName)
+	switch cloudName {
+	case "aws":
+		var awscfg aws.Config
+		awscfg.AccessKeyId = getEnvOrDefault("AWS_ACCESS_KEY_ID")
+		awscfg.SecretKey = getEnvOrDefault("AWS_SECRET_ACCESS_KEY")
+		provider, _ = aws.NewProvider(&awscfg)
+	case "libvirt":
+		var libvirtcfg libvirt.Config
+		libvirtcfg.URI = getEnvOrDefault("LIBVIRT_URI")
+		libvirtcfg.PoolName = getEnvOrDefault("LIBVIRT_POOL", "default")
+		libvirtcfg.NetworkName = getEnvOrDefault("LIBVIRT_NET", "default")
+		provider, _ = libvirt.NewProvider(&libvirtcfg)
+	default:
+		return fmt.Errorf("unknown provider: %s", cloudName)
+	}
+
+	if cloud := cloudmgr.Get(cloudName); cloud != nil {
+		cloud.LoadEnv() // this doesn't work ATM as it requires providers build tags for cloudmgr
+	}
```
If we can specify build tags here, I think we can simplify these lines to something like this.

```
	cloud := cloudmgr.Get(cloudName)
	if cloud == nil {
		return fmt.Errorf("unknown provider: %s", cloudName)
	}
	provider, err := cloud.NewProvider()
	if err != nil {
		return err
	}
```

snir911 replied:
I couldn't find how to do it, IIRC adding tags caused the controller-gen to fail
I thought we may be able to go the other way around, i.e. to build all unless there's a tag

snir911 replied:
done

yoheiueda commented on peer-pod-controller/Dockerfile:
```diff
@@ -15,13 +16,11 @@ COPY api/ api/
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
+RUN CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
```
Can we specify build tags for cloud providers here in the same way of other binariers like `cloud-api-adaptor` and `cluster-provisioner`?

Build tags are derived from `BUILTIN_CLOUD_PROVIDERS` in the main Makefile.
https://github.com/confidential-containers/cloud-api-adaptor/blob/900b330e4502ba219abcd900bdf7c541539e2a1c/Makefile#L21-L26

snir911 replied:
here's tags might work I'll have a look

yoheiueda commented on peer-pod-controller/controllers/peerpod_controller.go:
```diff
@@ -0,0 +1,148 @@
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.setProvider(ctx, pp.Spec.CloudProvider); err != nil { // TODO: set only once?
+		return ctrl.Result{}, err
+	}
```
If we only need to support a single cloud provider in a Kubernetes cluster, we don't need to set a cloud provider per pod here.

We can specify a cloud provider name as an argument to the controller binary, and create a Provider obect in `mainl.go` like [`cmd/cloud-api-adaptor/main.go`](https://github.com/confidential-containers/cloud-api-adaptor/blob/900b330e4502ba219abcd900bdf7c541539e2a1c/cmd/cloud-api-adaptor/main.go). Then we can pass the Provider object to this controller.

snir911 replied:
I used CLOUD_PROVIDER instead of setting as parameter to the binary

yoheiueda commented on peer-pod-controller/main.go:
```diff
@@ -84,6 +85,13 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "PeerPodConfig")
 		os.Exit(1)
 	}
+	if err = (&controllers.PeerPodReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "PeerPod")
+		os.Exit(1)
+	}
```
As mentioned above, if we can specify a cloud provider name as a command line argument, we can create a Provider object here, and pass it in this struct.

yoheiueda commented on pkg/util/k8sops/peerpod.go:
I think this package is only used in `"pkg/adaptor/cloud"`, so we can put this package under `"pkg/adaptor"`

yoheiueda commented on peer-pod-controller/controllers/peerpod_controller.go:
```diff
@@ -0,0 +1,148 @@
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("deleting instance", "InstanceID", pp.Spec.InstanceID, "CloudProvider", pp.Spec.CloudProvider)
+	if err := provider.DeleteInstance(ctx, pp.Spec.InstanceID); err != nil {
+		return ctrl.Result{}, err
+	}
```
Is `DeleteInstance` called by both controller and `cloud-api-adaptor`? If so, I think DeleteInstance will return an error like `instance not found`. Such an error probably means that the instance has been already deleted. Is it harmless to return such error here?

snir911 replied:
It seems the returned error comes from the provider's api, I'l check for the existence of the finalizer before deleting so that also if the deletion will fail due to `instance not found`, the next try will skip it and complete with success

yoheiueda commented on pkg/util/k8sops/peerpod.go:
```diff
@@ -0,0 +1,125 @@
+func NewPeerPodService() *PeerPodService {
+	cloudProvider := os.Getenv("CLOUD_PROVIDER") // TODO: don't get from env var directly
+	if cloudProvider == "" {
+		logger.Printf("NewPeerPodService: failed to get cloudProvider")
+		return nil
```
Can we return an error explicitly like this?

```
		return nil, errors.New("NewPeerPodService: failed to get cloudProvider")
```

yoheiueda commented on pkg/adaptor/cloud/cloud.go:
```diff
@@ -124,6 +126,9 @@ func NewService(provider Provider, proxyFactory proxy.Factory, workerNode podnet
 		workerNode:   workerNode,
 	}
 	s.cond = sync.NewCond(&s.mutex)
+	if s.ppService = k8sops.NewPeerPodService(); s.ppService == nil {
+		logger.Printf("failed to create PeerPodService, runtime failure may result in dangling resources")
+	}
```
Can we handle errors explicitly like this?

```
if ppService, err  := k8sops.NewPeerPodService(); err != nil {
	logger.Printf("failed to create PeerPodService, runtime failure may result in dangling resources: %v", err)
} else {
	s.ppService = ppService
}
```

yoheiueda reviewed (commented):
Thank you very much for proposing this great enhancement! This looks a right approach to me.

snir911 commented on peer-pod-controller/api/v1alpha1/peerpod_types.go:
```diff
@@ -0,0 +1,67 @@
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of PeerPod. Edit peerpod_types.go to remove/update
+	CloudProvider string `json:"cloudProvider,omitempty"`
```
Not really, at first i used it as it was not very simple to pass parameters at build time using the operator-sdk.
I'm still not sure how to do it but it indeed will be useful if we'll support multi cloud, and generally it just make sense to have context about the which provider the InstanceID refers to

yoheiueda replied:
OK. If we want to support multiple cloud providers, there is also possibility to support multiple accounts of a single cloud provider. So, we need a label to represents a pair of cloud provider name and account identifier.

A possible simple way to support such situation is to run multiple controllers, each of which is responsible only for a single cloud provider and a single account.  A controller should have a startup option to specify a value to be used for the label. The default value can be the same string of a cloud provider, since a single account is usually used for a single cloud provider.

Each controller determines which PeerPod objects it should handle by filtering the objects with the label.

yoheiueda replied:
I think is OK to keep this field. We will come back when we support multiple cloud providers.

stevenhorsman commented on go.mod:
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/IBM-Cloud/bluemix-go v0.0.0-20230120122421-afb48116b8f1
 	github.com/IBM/go-sdk-core/v5 v5.10.2
-	github.com/IBM/vpc-go-sdk v0.32.0
+	github.com/IBM/vpc-go-sdk v1.0.1
```
I think QiFeng purposely downgraded the vpc-go-sdk to the latest maintained one: https://github.com/confidential-containers/cloud-api-adaptor/pull/701#issuecomment-1463371224

snir911 commented:
I addresses all the issues commented, i hit some module issues due to the pkgs movements, hence, it blocks me from testing ATM but i don't expect code to have additional dramatic changes.
(also need to rename peer-pod-ctlr to peer-pod-ctrl)

snir911 commented:
removed the WIP & RFC, please review
**NOTE:** it still requires to remove some hacks that was done for testing before merging
all other changes can be done as follow up also supporting other providers

bpradipt commented on peer-pod-ctrl/controllers/peerpod_controller.go:
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2023.
```
nit: Better to use the confidential containers copyright

yoheiueda commented:
> Please remove any commits that are really not required for peerpod CR. For example I think this commit is not needed - all: mv cloudmgr pkg under pkg/

I think this specific change is based on my previous comment.

It is OK with me to keep using the original package location of `cmd/cloud-api-adaptor/cloudmgr`. I'll submit another PR to reorganize the `cloudmgr` functionality into each cloud provider code in `pkg/cloud/*`.

bpradipt commented:
> > Please remove any commits that are really not required for peerpod CR. For example I think this commit is not needed - all: mv cloudmgr pkg under pkg/
> 
> I think this specific change is based on my previous comment.
> 
> It is OK with me to keep using the original package location of `cmd/cloud-api-adaptor/cloudmgr`. I'll submit another PR to reorganize the `cloudmgr` functionality into each cloud provider code in `pkg/cloud/*`.

@yoheiueda thanks for clarifying. I think since this PR is quite big, keeping it specific to the controller required changes will be better from review and test standpoint. wdyt ?

snir911 commented:
> @snir911 thanks for implementing this important feature. Overall it looks good. I need to do a second pass but before that few generic comments to help move forward.
> 
> 1. Please squash few commits. For example the reconcile loop function and the api etc creation can be in one PR.

you mean to split the api calles from the CAA to a different PR?

> 2. Please remove any commits that are really not required for peerpod CR. For example I think this commit is not needed - `all: mv cloudmgr pkg under pkg/`

As @yoheiueda mentioned i moved it as he requested, however, i think reverting it and reorganize later also make sense, also i think it would help to avoid the need to point go.mod to the root dir
 
> 3. Add proper copyright statements for CoCo

ok
> 4. Please rename the folder to `peerpod-controller` to match the CR - `peerpod`.

The reason i used "ctrl" and not "controller" is because when i tried to stick with this naming in the other operator ([peer-pod-config-controller](https://github.com/confidential-containers/cloud-api-adaptor/pull/714) it resulted in failure with some of the objects definitions due to a too long name (default names the operator-sdk is using)
Of course it can be modified, but i thought it would be better to be unified with the naming conventions in both operator

bpradipt commented:
> > @snir911 thanks for implementing this important feature. Overall it looks good. I need to do a second pass but before that few generic comments to help move forward.
> > 
> > 1. Please squash few commits. For example the reconcile loop function and the api etc creation can be in one PR.
> 
> you mean to split the api calles from the CAA to a different PR?

These two commits
- 8e0206b11a9e3d73a9090ea08592e379a7e59553
- b2b243461f9ebc4851cdf9855c3e0f4d48ab801e
- 
> 
> > 4. Please rename the folder to `peerpod-controller` to match the CR - `peerpod`.
> 
> The reason i used "ctrl" and not "controller" is because when i tried to stick with this naming in the other operator ([peer-pod-config-controller](https://github.com/confidential-containers/cloud-api-adaptor/pull/714) it resulted in failure with some of the objects definitions due to a too long name (default names the operator-sdk is using) Of course it can be modified, but i thought it would be better to be unified with the naming conventions in both operator

:-( ..Let's use `peerpod-ctrl` (peerpod as a single word to denote the CR)

bpradipt reviewed (approved):
/lgtm
Thanks @snir911. The best part is that the actual controller code is pretty less :)

yoheiueda reviewed (approved):
LGTM
